util/merkle-mountain-range/src/helper.rs

```diff
@@ -0,0 +1,110 @@
+pub fn leaf_index_to_pos(index: u64) -> u64 {
+    if index == 0 {
+        return 0;
+    }
+    // leaf_count
+    let mut leaves = index + 1;
+    let mut tree_node_count = 0;
+    let mut height = 0u32;
+    while leaves > 1 {
+        // get heighest peak height
+        height = (leaves as f64).log2() as u32;
+        // calculate leaves in peak
+        let peak_leaves = 1 << height;
+        // heighest positon
+        let sub_tree_node_count = get_peak_pos_by_height(height) + 1;
+        tree_node_count += sub_tree_node_count;
+        leaves -= peak_leaves;
+    }
+    if leaves > 0 {
+        // // add one pos for remain leaf
+        // let pos = tree_node_count - 1;
+        // pos + 1
+        tree_node_count
+    } else {
+        let pos = tree_node_count - 1;
+        pos - u64::from(height)
+    }
+}
+
+// TODO optimize
+pub fn leaf_index_to_mmr_size(index: u64) -> u64 {
+    let mut pos = leaf_index_to_pos(index);
+    while pos_height_in_tree(pos + 1) > pos_height_in_tree(pos) {
+        pos += 1
+    }
+    pos + 1
+}
+
+pub fn pos_height_in_tree(mut pos: u64) -> u64 {
+    pos += 1;
+    fn all_ones(num: u64) -> bool {
+        num.count_zeros() == num.leading_zeros()
```

Comment:
```
let num = 0u8;
assert_eq!(num.count_zeros(), num.leading_zeros());
```

I think `0u8` should not be `all_ones`.

Reply:
But it's not a big problem since the index begins with 1.

Reply:
Good catch